### DISCUSSION TITLE
Are the NvKmsKapiFunctionTable functions single-buffered or double-buffered?

### DISCUSSION THREAD

**Discussion by Cacodemon345** [NVIDIA/open-gpu-kernel-modules#189] [Q&A] [answered]
As the title says.

Looking at https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/kernel-open/common/inc/nvkms-kapi.h doesn't tell me much about whether it's single buffered or double buffered.

**Comment by aritger** [comment]
The NvKmsKapiFunctionsTable structure is assigned by nvKmsKapiGetFunctionsTableInternal(), overwriting any existing values in that structure, if I understand the question correctly.

**Reply by Cacodemon345** [reply]
What I am saying is, can the functions work on a single framebuffer surface without the need for flipping?

**Reply by misyltoad** [reply]
I have no context of the NVIDIA specific interfaces, but for DRM atomic, the fb_ids (surface)s you provide will be used for the next scanout.

**Comment by aritger** [comment, marked as answer]
The functions in nvkms-kapi.h are part of nvidia-drm.ko's DRM atomic implementation, so Joshua-Ashton's response applies.  That said, yes, you can work on the currently displayed surface without flipping if you really want, with the usual artifacts that will create.

**Reply by Cacodemon345** [reply]
One more thing: Does releasing ownership cause all surfaces created while grabbed to be made invalid?

**Reply by aaronp24** [reply]
Releasing ownership explicitly doesn't, but closing the file descriptor does. See the call chain from nvKmsClose -> FreeDeviceReference -> [nvEvoFreeClientSurfaces](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/515.43.04/src/nvidia-modeset/src/nvkms-surface.c#L895).

**Reply by Cacodemon345** [reply]
Is it possible to create a surface not tied to any file descriptors then?

**Reply by aaronp24** [reply]
Can you describe your use case in a little more detail? It might help try to figure out what the right approach is.

Surface registrations are associated with a [`struct NvKmsPerOpenDev`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/515.43.04/src/nvidia-modeset/src/nvkms.c#L65) object. These are normally associated with open `/dev/nvidia-modeset` file descriptors but there are a couple of exceptions:
 * [`nvRmImportFbConsoleMemory`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/515.43.04/src/nvidia-modeset/src/nvkms-rm.c#L4601) uses one that belongs to NVKMS itself to track the memory holding the graphical framebuffer console, if one exists
 * The [`CreateSurface`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/515.43.04/src/nvidia-modeset/kapi/src/nvkms-kapi.c#L1917) function in the kapi layer uses one to register NVKMS surfaces on behalf of nvidia-drm.ko.

**Reply by Cacodemon345** [reply]
I would love to do fbdev support so that Linux is able to select the highest framebuffer resolution at boot time instead of being stuck with the boot-time bootloader resolution.

**Reply by aaronp24** [reply]
Ah, okay. In that case, take a look at how nvidia-drm calls into nvidia-modeset via the kapi interface.

That being said, please be aware that one long-standing feature that we support is the ability to unload the NVIDIA drivers, install a different version, and then start everything back up. I believe that doesn't quite work today with this particular version of the driver but it's a feature we technically still need to support today. If plugging in fbdev support prevents the driver from being unloaded then I think it'll have to be behind an option that's disabled by default, at least for now.

**Reply by Cacodemon345** [reply]
The driver doesn't look to be designed for fbdev though, which is so far my impression.

I would prefer fbdev support to be developed by others or NVIDIA themselves, since I am not that experienced in GPU/display driver development.

**Reply by hramrach** [reply]
Why is a fbdev interface needed for modesetting? Can't KMS be already used?

**Reply by Cacodemon345** [reply]
Because the Linux kernel relies on fbdev for its framebuffer console. NVIDIA drivers do not provide both KMS and fbdev, only the former, so only vesa/efifb works with potentially slow speed. AMD and Intel drivers on Linux provide both, so the console isn't stuck on a low resolution and can render the console efficiently.

This has been a problem on FreeBSD's side as well, where framebuffer support on NVIDIA is problematic.

**Reply by hramrach** [reply]
You can probably use something like plymouth to display kernel messages using KMS, and kmscon to run full featured graphics console on KMS (as opposed to the stub implementation provided by the kernel).

**Reply by Cacodemon345** [reply]
Not everyone uses kmscon. Many may even choose to stay away from it with being unwilling to deal with issues starting Wayland compositors or Xorg.

Plymouth still wouldn't solve the low resolution problem.

**Reply by hramrach** [reply]
Why it would not solve low resolution problem? It does have a KMS driver.

**Reply by hramrach** [reply]
And what issues starting Wayland or Xorg are there? All issues starting Wayland and Xorg that are not outright problem with the driver or the display server are caused by fbdev. Without fbdev what issues are there?

**Reply by Cacodemon345** [reply]
> Why it would not solve low resolution problem? It does have a KMS driver.

Because only the graphical screen would run in high resolution. Once the console reappears you would be back in the old fbdev resolution.

**Reply by Cacodemon345** [reply]
> And what issues starting Wayland or Xorg are there? All issues starting Wayland and Xorg that are not outright problem with the driver or the display server are caused by fbdev. Without fbdev what issues are there?

DRM master issues with kmscon if seatd isn't running and the init system is anything but systemd.

If you run from the Linux framebuffer console you don't run into such issues at all.

**Reply by hramrach** [reply]
If you don't have fbdev there is no fbdev resolution so you can't get any fbdev resolution when console reappears. If you are using kmscon then you get console in graphics.

If you are using fbdev then you have another problem - any program that opens the fb blocks the graphics completely, and this is not possible to solve with any kind of arbitration, unlike the DRM master problem.

**Reply by Cacodemon345** [reply]
But the problem is that on average x86 desktops and workstations, Linux will tend to stick with bootloader's framebuffer with either vesa and efifb at initial stage before loading KMS drivers, and so far only built-in KMS drivers provide both KMS and fbdev and therefore are able to switch the resolution to the highest one when loading those.

KMS alone isn't enough to switch framebuffer console resolutions; it only provides DRI interfaces for Xorg/Wayland to use it. When you switch from a Xorg/Wayland screen to a framebuffer console VT screen, you will see that the monitor can blank out when both are using different resolutions. Which means KMS can't be used to set framebuffer console resolution on other terminal VTs.

The official NVIDIA drivers only deal with Xorg/Wayland DRI interfaces and does not deal with fbdev, so Linux will be still stuck with vesa/efifb because it provides no fbdev support for the kernel to use for its framebuffer console, whereas the built-in GPU/display drivers provide fbdev support for the kernel to use for high-resolution framebuffer consoles.

**Reply by hramrach** [reply]
If you do care about high resolution console kmscon and similar do provide that on top of KMS. 

vesafb/efifb cannot be active together with a KMS driver so when nVidia driver loads these should be deactivated.

fbdev is broken and can never work so it would be backwards for the nVidia drivers to implement it.

Instead it would be more constructive to add support for KMS to init systems other than systemd if you care about that.

**Reply by Cacodemon345** [reply]
Except kmscon isn't officially maintained anymore; only unofficial versions. And not many people know or even bother to care about it either, instead being content with the framebuffer console provided by Linux.

And generally other init systems follow the Unix philosophy in doing a single thing and doing so well which again leaves KMS to be dealt with by other programs and not the init system itself.

**Reply by Cacodemon345** [reply]
For reference, see https://github.com/NVIDIA/open-gpu-kernel-modules/issues/228 where Fedora's removal of fbdev drivers left NVIDIA users with non-working framebuffer console output, because again, NVIDIA drivers do not implement fbdev. There also isn't any effort to port the Linux framebuffer console to work with KMS drivers that doesn't provide fbdev support to my understanding.
